### PR TITLE
Docs for process for refactoring a Primer React Component to CSS modules

### PR DESCRIPTION
**Pull request by jonrohan**
Closes https://github.com/github/primer/issues/3657

This PR documents how to migrate a styled component into a CSS module in primer react.

### CODE REVIEW

File: contributor-docs/migrating-to-css-modules.md [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+1. **Check for `className` Prop:**
+   - Ensure the component accepts a `className`.
+2. **Feature Flagging:**
+   - Add a feature flag to toggle the `sx` prop for controlled rollout (staff shipping).
+3. **Create CSS Module:**
+   - Add a corresponding `{Component}.module.css` file.
+4. **Import CSS Modules:**
+   - Use the CSS module in the component.
+   - Add CSS classes behind the `primer_react_css_modules_team` feature flag.
+5. **Ensure Component still accepts sx styling**
```
Comment (joshblack) [inline contributor-docs/migrating-to-css-modules.md:50]: Unsure of how to word this, but may be helpful to include in addition to point 1

```suggestion
5. **Ensure Component still accepts sx styling**
6. **Ensure Component accepts `className` along with our CSS Module class name**
```

File: contributor-docs/migrating-to-css-modules.md [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+### Refactoring Styled-Components to CSS Modules
+
+- **Replace `${get('...')}` Syntax:**
+  - Migrate these to CSS Modules classes.
+  - No need for fallbacks in CSS Modules.
+
+### When Refactoring a Component
+
+1. **Check for `className` Prop:**
+   - Ensure the component accepts a `className`.
```
Comment (iansan5653) [inline contributor-docs/migrating-to-css-modules.md:41]: Should migrated components also accept an inline `style` prop? This is sometimes necessary for very dynamic styling (ie, style values that depend on props).

File: contributor-docs/migrating-to-css-modules.md [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+2. **Feature Flagging:**
+   - Add a feature flag to toggle the `sx` prop for controlled rollout (staff shipping).
+3. **Create CSS Module:**
+   - Add a corresponding `{Component}.module.css` file.
+4. **Import CSS Modules:**
+   - Use the CSS module in the component.
+   - Add CSS classes behind the `primer_react_css_modules_team` feature flag.
+5. **Ensure Component still accepts sx styling**
+   - Until we migrate all uses of sx, we need to ensure the component will accept sx props inside the feature flag. This will often default to using the `Box` component if an sx prop is passed in.
```
Comment (joshblack) [inline contributor-docs/migrating-to-css-modules.md:51]: Might be a good spot for a snippet to include that we use `Box` for this (and potentially `toggleStyledComponent` if we want that in here)

File: contributor-docs/migrating-to-css-modules.md [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+### Testing the Migration
+
+- **Support for `className`:**
+  - Ensure the component works properly with the `className` prop.
```
Comment (joshblack) [inline contributor-docs/migrating-to-css-modules.md:56]: Good place for a code snippet to drop-in, could help to demonstrate

- What the test would look like
- How to use feature flag component

File: contributor-docs/migrating-to-css-modules.md [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+- **Regression Testing:**
+  - Validate that no regressions occur when the feature flag is enabled.
```
Comment (joshblack) [inline contributor-docs/migrating-to-css-modules.md:58]: Might help to provide more context here that we are looking for no visual regressions in our snapshots (so vrt check s hould not fail)

File: contributor-docs/migrating-to-css-modules.md [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+- **Handling `sx` Prop:**
+  - Confirm the `sx` prop behaves correctly with the feature flag enabled.
```
Comment (joshblack) [inline contributor-docs/migrating-to-css-modules.md:60]: I think we use two places to potentially test for this:

- Jest test to confirm `sx` is being passed
- Dev story (maybe?)

Curious what you would say 👀

File: contributor-docs/migrating-to-css-modules.md [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+
+---
+
+## Releasing a Component
+
+- **Merge Process:**
+  - Once merged into dotcom, the component will be part of the `primer_react_css_modules_team` rollout.
+  - We progress components through levels of feature flags as we roll them out to make sure we address any bugs with the minimum amount of audience.
+- **Monitor Team Ship:**
+  - Watch for any issues or regressions during the team ship phase.
```
Comment (joshblack) [inline contributor-docs/migrating-to-css-modules.md:70]: Would be helpful to call out where to flag regressions (in slack) and how to capture them in issues for us to review (have a process to pitch for this in our engineering session this week)

File: contributor-docs/migrating-to-css-modules.md [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+  - No need for fallbacks in CSS Modules.
+
+### When Refactoring a Component
+
+1. **Check for `className` Prop:**
+   - Ensure the component accepts a `className`.
+2. **Feature Flagging:**
+   - Add a feature flag to toggle the `sx` prop for controlled rollout (staff shipping).
+3. **Create CSS Module:**
+   - Add a corresponding `{Component}.module.css` file.
```
Comment (joshblack) [inline contributor-docs/migrating-to-css-modules.md:46]: Would this be a good place to talk about how we use `:where()` for managing specificity or would this be a different doc?

**Review by joshblack — COMMENTED** [review]
Looking great! Left a couple of comments 👀

File: contributor-docs/migrating-to-css-modules.md [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+    - [Testing the Migration](#testing-the-migration)
+  - [Releasing a Component](#releasing-a-component)
+
+---
+
+## Before Refactoring a Component
+
+- **Verify VRT (Visual Regression Testing) Coverage:**
+  - Check for missing VRT coverage. We utilize the VRT tests to make sure we're matching styling in production with current expectations.
+  - Make sure there are `dev` stories for any edge cases spotted in production for the component (ie. sx prop, custom className, styled system attributes).
```
Comment (langermank) [inline contributor-docs/migrating-to-css-modules.md:24]:
```suggestion
  - Make sure there are `dev` stories for any edge cases spotted in production for the component (ie. sx prop, custom className, styled system attributes). `dev` stories may include things that we wouldn't normally recommend for the purpose of stress testing what happens when PRC components are overridden with custom styles. 
```

File: contributor-docs/migrating-to-css-modules.md [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+  - Make necessary visual changes **before** creating the CSS Modules refactor PR.
+
+---
+
+## How to Migrate
+
+### Refactoring Styled-Components to CSS Modules
+
+- **Replace `${get('...')}` Syntax:**
+  - Migrate these to CSS Modules classes.
```
Comment (langermank) [inline contributor-docs/migrating-to-css-modules.md:35]: Do you mean CSS variables here? I would link to https://primer.style/foundations/primitives/color and then include a code example showing the conversion

File: contributor-docs/migrating-to-css-modules.md [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+  - No need for fallbacks in CSS Modules.
+
+### When Refactoring a Component
+
+1. **Check for `className` and `style` Prop:**
+   - Ensure the component accepts a `className` for styling from outside of primer/react.
```
Comment (langermank) [inline contributor-docs/migrating-to-css-modules.md:41]:
```suggestion
   - Ensure the component accepts a `className` *on the top dom level only* for styling from outside of primer/react.
```

File: contributor-docs/migrating-to-css-modules.md [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+   - Ensure the component accepts a `style` prop for more dynamic styling like positioning.
```
Comment (langermank) [inline contributor-docs/migrating-to-css-modules.md:42]: Do we have to explicitly set this as a prop, or does it just come for free? CC @joshblack

File: contributor-docs/migrating-to-css-modules.md (hunk 8 above)
Comment (langermank) [reply contributor-docs/migrating-to-css-modules.md:46]: Maybe we could just link to the doc I'm writing here when its done

Comment (jonrohan) [reply contributor-docs/migrating-to-css-modules.md:46]: I think that would fall into the css authoring doc https://github.com/primer/react/pull/5037/files

File: contributor-docs/migrating-to-css-modules.md [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+2. **Feature Flagging:**
+   - Add a feature flag to toggle the `sx` prop for controlled rollout (staff shipping).
+3. **Create CSS Module:**
+   - Add a corresponding `{Component}.module.css` file.
+4. **Import CSS Modules:**
+   - Use the CSS module in the component.
```
Comment (langermank) [inline contributor-docs/migrating-to-css-modules.md:48]: I would include a code chunk here showing what the import looks like

File: contributor-docs/migrating-to-css-modules.md [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+   - Add CSS classes behind the `primer_react_css_modules_team` feature flag.
```
Comment (langermank) [inline contributor-docs/migrating-to-css-modules.md:49]: I think this is lacking a lot of context. We have been using different syntax for different scenarios, do you think we should document each of those scenarios and show a small snippet of what the conversion looks like? Like if X syntax, use Y syntax to feature flag etc etc

**Review by langermank — APPROVED** [review]
Great start! Just left some ideas

File: contributor-docs/migrating-to-css-modules.md [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+  - Migrate these to CSS Modules classes.
+  - No need for fallbacks in CSS Modules.
+
+### When Refactoring a Component
+
+1. **Check for `className` and `style` Prop:**
+   - Ensure the component accepts a `className` *on the top dom level only* for styling from outside of primer/react.
+   - Ensure the component accepts a `style` prop for more dynamic styling like positioning.
+2. **Feature Flagging:**
+   - Add a feature flag to toggle the `sx` prop for controlled rollout (staff shipping).
```
Comment (randall-krauskopf) [inline contributor-docs/migrating-to-css-modules.md:44]: I think it might be helpful to add a snippet showing what the feature flag code looks like for toggling the sx prop. Or even link to a previous PR that shows what a total migration might look like.

File: contributor-docs/migrating-to-css-modules.md (hunk 12 above)
Comment (joshblack) [reply contributor-docs/migrating-to-css-modules.md:42]: @langermank it will come for free with `{...props}` most of the time but it may be good for us to make both explicit to avoid issues like:

```tsx
<MyComponent
  className="hi"
  styles={{ /* ... */ }}
  {...props}
/>
```

Where either could be overridden instead of merged like we would want

**Review by langermank — APPROVED** [review]
Fantastic!
